044-kraft-liveness-readiness.md

```diff
@@ -0,0 +1,178 @@
+# Liveness and Readiness probes in a KRaft Kafka cluster
+
+This proposal describes the liveness and readiness probes that should be put in place 
+for a Kafka cluster that is using KRaft rather than ZooKeeper. This includes a KRaft 
+cluster with combined nodes and one that has separate controller and broker nodes.
+
+## Current situation
+
+These are the current probes used in Strimzi when ZooKeeper is present:
+
+|Component| Liveness | Readiness |
+|---|---|---|
+| ZooKeeper | "imok" from 127.0.0.1:12181 | "imok" from 127.0.0.1:12181 |
+| ZK-based broker | if (ready) { listening on 9091 } else { have ZK session } | if (ready) { return true } |
+
+The "ready" check in both the liveness and readiness checks for a Zk-based broker is based on the existence of the file `../kafka-ready` which is 
+created by the KafkaAgent when the broker state metric has value 3 (RUNNING).
+
+## Background
+
+This section contains some background information about the ports and metrics we could use for checks.
+
+### Ports
+
+In KRaft mode there are two different ports that the node could be listening on.
+If it has a controller process it should be listening on the port defined as the `controller.listener.names` which in Strimzi is hard-coded to 9090.
+If it has a broker process it should be listening on the replication port specified as the `inter.broker.listener.name` which is hard-coded to 9091.
+If the node is running as a combined mode, it should be listening on both ports
+
+### BrokerState
+
+The expected [BrokerState](https://github.com/apache/kafka/blob/trunk/metadata/src/main/java/org/apache/kafka/metadata/BrokerState.java) transitions in the "happy path" during a normal broker lifecycle are:
+
+NOT_RUNNING (0) -> STARTING (1) -> RECOVERY (2) -> RUNNING (3) -> PENDING_CONTROLLED_SHUTDOWN (6) -> SHUTTING_DOWN (7)
+
+The other possible value is 127 which represents UNKNOWN.
+
+You can see how the state transitions in the [BrokerLifecycleManager](https://github.com/apache/kafka/blob/trunk/core/src/main/scala/kafka/server/BrokerLifecycleManager.scala#L379), specifically that 
+the broker transitions from STARTING to RECOVERY once it has caught up with the cluster metadata, then it transitions to RUNNING as long as it is not fenced.
+
+The broker nodes cannot catch up with the cluster metadata until the controller quorum has been formed, so they will not move out of the STARTING state until the quorum leader election has happened.
+
+## Proposal
+
+There are three different types of liveness/readiness probe that we need to consider, first for a controller only, then 
+for a broker only, and finally for a combined node that has both roles.
+
+The following statements describe the intent of the proposed probes:
+
+* A controller is:
+    * "alive" if it has a process running.
+    * "ready" when it is ready to join the controller quorum.
```

Comment:
just to be sure I got this right. Isn't "ready" when it joined a controller quorum? or actually when it's ready to join but not joined yet? I would expect the former, or? Maybe just a matter of how it sounds in English vs Italian? :-)

Reply:
Ok, makes sense now, thanks.